### PR TITLE
Avoid accessing vectors with zero length, by checking data sizes.

### PR DESCRIPTION
**Pull request by larshg**

### CODE REVIEW

**Comment by larshg** [comment]
Errors appeared when running IO Unittests in debug mode, hence not detected by CI.

File: io/include/pcl/io/pcd_io.h [hunk 1]
```diff
@@ -281,9 +281,20 @@ namespace pcl
         int res = read (file_name, blob, cloud.sensor_origin_, cloud.sensor_orientation_,
                         pcd_version, offset);
 
-        // If no error, convert the data
-        if (res == 0)
-          pcl::fromPCLPointCloud2 (blob, cloud);
+        // If no error, convert the data if there is data
```
[the remark this answers is not in the document]
Comment (larshg) [reply io/include/pcl/io/pcd_io.h:284]: Indeed it could :)

File: io/src/pcd_io.cpp [hunk 2]
```diff
@@ -342,6 +342,9 @@ pcl::PCDReader::readHeader (std::istream &fs, pcl::PCLPointCloud2 &cloud,
   if (nr_points == 0)
   {
     PCL_WARN ("[pcl::PCDReader::readHeader] No points to read\n");
+    cloud.width = 0;
+    cloud.data.clear();
+    return 0;
```
[the remark this answers is not in the document]
Comment (larshg) [reply io/src/pcd_io.cpp]: No, I could remove this check entirely 😅

File: common/include/pcl/conversions.h [hunk 3]
```diff
@@ -169,6 +169,13 @@ namespace pcl
     cloud.height   = msg.height;
     cloud.is_dense = msg.is_dense == 1;
 
+    // check if there is data to copy
+    if (msg.data.empty())
```
Comment (mvieth) [inline common/include/pcl/conversions.h:173]: `msg.data` should not be used in this function, but `msg_data`.
However, I would say `msg.width*msg.height == 0` is the best check to determine whether `cloud[0]` causes a problem, no?

File: io/src/pcd_io.cpp [hunk 4]
```diff
@@ -339,11 +339,6 @@ pcl::PCDReader::readHeader (std::istream &fs, pcl::PCLPointCloud2 &cloud,
                                     [](const pcl::PCLPointField& field)->bool { return field.count < 1; }),
                      cloud.fields.end());
 
-  if (nr_points == 0)
-  {
-    PCL_WARN ("[pcl::PCDReader::readHeader] No points to read\n");
```
Comment (mvieth) [inline io/src/pcd_io.cpp:344]: Let's maybe keep the warning?

**Review by mvieth — APPROVED** [review]

**Pull request merged**